### ISSUE TITLE
Check if a file or directory exists

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#6713]
## Problem

The Dagger API does not support checking if a file exists. There is an SDK-specific workaround (maybe? not documented) which is to 1) call any function that requires the file to exist, like `Sync` or `Contents`, 2) check for error, and 3) check for the error message returned when the file doesn't exist. This possibly can be wrapped in a convenience type by the SDK (again hard to tell as it is not documented).

This makes for a bad DX.

## Solution

Several solutions have been proposed:

### Option 1: `File.exists()`, `Directory.exists()`

Add a new core function to the `File` and `Directory` types:

```graphql
type Directory {
 "Returns true if the directory exists"
 exists: Bool!
}
 
type File {
 "Returns true if the file exists"
 exists: Bool!
}
```

### Option 2: `Directory.exists(path)`

In this option, you query a parent directory for the existence of a given path:

```graphql
type Directory {
 "Returns true if a file or directory exists at the given path"
 exists(path: String!): Bool!
}
```

**Comment by helderco** [comment]
I think there’s an issue for this already. Maybe via a Stat field.

**Comment by kpenfound** [comment]
This should also only return an error if something actually went wrong. If the file/directory doesn't exist, this should return `False` and not error

Maybe stating the obvious, but for the sake of being super clear 😂

**Comment by kpenfound** [comment]
Related (maybe dupe, maybe not): https://github.com/dagger/dagger/issues/3764

**Comment by helderco** [comment]
What I do is call `dir.Entries(ctx)` and then check if the file is in the list.

**Comment by helderco** [comment]
> Related (maybe dupe, maybe not): #3764

`Stat` would have to return an error if a file doesn't exist (custom `NotExistError`, like `ExecError`). We can definitely do that, but also cool with an `.Exists()` convenience that returns a `bool`.

**Comment by jalvarezit** [comment]
> > Related (maybe dupe, maybe not): #3764
> 
> `Stat` would have to return an error if a file doesn't exist (custom `NotExistError`, like `ExecError`). We can definitely do that, but also cool with an `.Exists()` convenience that returns a `bool`.

That would fit definetely, also as mentioned in #3764 the required function is already in the [core library](https://github.com/dagger/dagger/blob/ab38c29af3283134f2ec43301f51585083bfb9a9/core/directory.go#L182).

Not sure how difficult is to ship that method to the sdk so we can reuse it.

For instance I had to implement a workaround to check if a path is a file or not just because I do not have stat functionality available:

```go
func (m *DaggerStructureTest) IsDir(ctx context.Context, container *Container, path string) (bool, error) {

	_, isNotDirErr := container.Directory(path).Sync(ctx)

	// If it is a directory do not keep further checking
	if isNotDirErr == nil {
		return true, nil
	}

	_, isNotFileErr := container.File(path).Sync(ctx)

	if isNotFileErr == nil {
		return false, nil
	}

	// At this point we know that the path does not exist or a graphql error occurred
	// We also assume that isNotDirErr and isNotFileErr are the same error
	return false, isNotFileErr
}
```

**Comment by helderco** [comment]
Adding an `exists` field to `Directory` doesn't work without a `path`. Consider this:

```golang
exists, err := ctr.Directory(path).Exists(ctx)
```

These fields already validate if the directory/file exists and if it's a directory/file:

* `Container.directory`
* `Directory.directory`
* `Container.file`
* `Directory.file`

So the above snippet fails on `ctr.Directory(path)`, before reaching `.Exists(ctx)`.

This behavior is appropriate because on most operations you want the early error, rather than a possibly more confusing one later due to an inexistent directory/file.

I suggest adding a `path` argument instead:

```graphql
type Container {
    exists(path: String!): Boolean!
}

type Directory {
    exists(path: String!): Boolean!
}
```

Then in code it becomes:

```diff
- exists, err := ctr.Directory(path).Exists(ctx)
+ exists, err := ctr.Exists(ctx, path)
```

However you can't tell if it's a directory or a file this way because `path` could be either, unless you have specific:

* `Container.existsDirectory`
* `Container.existsFile`
* `Directory.existsDirectory`
* `Directory.existsFile`

But I think I prefer relying on `[Stat](https://github.com/dagger/dagger/issues/3764)` for that instead of the above fields. We can also create a custom error like `dagger.ExecError` (e.g., `dagger.NotExistsError`) so `Stat` can serve both purposes, but that's best in a separate issue. Doesn't mean we don't do both.

We can add the following as well:

```graphql
type File {
    exists: Boolean!
}
```

But I'm not sure about its value since things that return a `File` usually validate if it exists already. We just need the custom error to check in code, rather than looking at the error message.

Does anyone know a use case where you have a `File` instance that may not exist? `dir.File("not-exists")` doesn't count because no method on that `File` will work since `Directory.file` will return an error before that.

**Comment by jalvarezit** [comment]
I definetely love the idea of having both Stat and Exists methods, they would complement each other perfectly. Code may end being a bit more verbose but it gives more fine grain control over the checks you want to pass. 

@helderco I can think of a possible scenario. We are using a container that renders some output files in a directory and then we pass that resulting directory to another container where we extract some of these rendered files. However, if there was some sort of error or bug in the rendering process may cause the file not existing in the directory. Would be great to be able to check if the file exists to allow some error handling instead of just panicking the process.

**Comment by shykes** [comment]
It makes more sense to me to resolve file/directory existence lazily, and keep `File.Exists()` and `Directory.Exists()` as proposed, without path. It's the most intuitive and doesn't break any existing use cases.

> This behavior is appropriate because on most operations you want the early error, rather than a possibly more confusing one later due to an inexistent directory/file.

This doesn't make sense to me. As a dev I call `ctr.Directory(foo)` without having to check for error, which tells me the directory object is lazy and may not exist. If you just add a `Exists` function to that lazy object, I can cleanly find out, and everyone is happy.

**Comment by shykes** [comment]
I'm against `Stat` because it's a posix relic that is unnecessary when you already have first class support for structured data in the core API. In a C syscall API you need a shorthand to check for existence + get a struct with various information about a filesystem entry all at once. But we don't have the constraints of C: we have  Directory a File types which can be queried as structs directly. We don't need the `stat` hack. Let it die in peace :)

**Comment by helderco** [comment]
> It makes more sense to me to resolve file/directory existence lazily, and keep `File.Exists()` and `Directory.Exists()` as proposed, without path. It's the most intuitive and doesn't break any existing use cases.

I knew using Go examples would be confusing. 🙂 I tried explaining that this happens while processing the pipeline server side, not on the client. I can make it clearer.

I agree that it would be more intuitive but when you do this:

```go
dir := dag.Container().From("alpine").Directory("not-exists")
// ...
exists, err := dir.Exists(ctx)
```

It means `Directory()` is indeed lazy, until you reach the leaf `Exists()`, which forces the query to be built and executed:

```graphql
query {
  container {
    from(address: "alpine") {
      directory(path: "not-exists") {
        exists
      }
    }
  }
}
```

At this point, the server resolves these steps sequentially: *container* → `directory` resolver → `exists` resolver.

<details><summary>🔀 Flowchart</summary>

```mermaid
flowchart TB
    subgraph D ["directory(path: #quot;not-exists#quot;)"]
        DE{Does it exist?}
        DE --> DEY[Yes]
        DE --> DEN[No]
        DEY -->DD{Is it a directory?}
        DD --> DDY[Yes]
        DD --> DDN[No]
    end
    subgraph E ["exists"]
        EE{Does it exist?} --> EEY[Yes]
        EE --> EEN[No]
    end
    C(container) --> D
    DEN & DDN --> RE[Return Error]
    DDY --> E
    EEY -->|true| R[Return boolean]
    EEN -.->|false ❌| R
```

</details>

As you can see in the flowchart, this validation already happens in `Container.directory` so the request [returns an error at that location](https://play.dagger.cloud/playground/Ierm9_DiU_m):

```json
{
  "errors": [
    {
      "message": "resolve: lstat /not-exists: no such file or directory",
      "path": [
        "container",
        "from",
        "directory"
      ]
    }
  ],
  "data": null
}
```


It doesn’t get the chance to reach the logic in the `exists` resolver. You certainly can’t catch the error from the `directory` resolver inside the `exists` resolver.

As I mentioned, this validation already happens in the following fields:

- `Container.directory`
- `Directory.directory`
- `Container.file`
- `Directory.file`

> > This behavior is appropriate because on most operations you want the early error, rather than a possibly more confusing one later due to an inexistent directory/file.
> 
> This doesn't make sense to me. As a dev I call `ctr.Directory(foo)` without having to check for error, which tells me the directory object is lazy and may not exist. If you just add a `Exists` function to that lazy object, I can cleanly find out, and everyone is happy.

By “this behavior is appropriate” I’m talking about the validation that happens in the `directory()` resolver. To make `exists` work the way you propose, I’d have to remove this validation from the fields above, but it was added for a reason: better error message in every use case beside `exists`.

Remember that we had a similar issue with `ExitCode`. It never returned a `false` value because it would exit early with an error. The solution there was to make a custom error that you can catch (`ExecError`).

However, we can keep the validation in those fields and still have an `exists`, but **you need to call it from a `Directory` that you know should exist, and pass it the path that you’re unsure about**.

**Comment by helderco** [comment]
> However, we can keep the validation in those fields and still have an `exists`, but **you need to call it from a `Directory` that you know should exist, and pass it the path that you’re unsure about**. Unfortunately this doesn't distinguish between files or directories unless you make specific `existsDirectory` and `existsFile` fields.

The other alternative is a custom error, as I said before:

```go

dir, err := dag.Container().From("alpine").Directory("not-exists").Sync(ctx)
var e *dagger.NotExistsError
if errors.As(err, &e) {
    // ...
}
```

We could make a convenience in the SDK (`dir.Exists(ctx)`), but it’s best if we can find a rule to encode it in codegen[^1] (like `sync`), potentially covering other similar kinds of errors, rather than being an isolated thing.

[^1]: With the [centralized codegen](https://github.com/dagger/dagger/issues/5226) work.

**Comment by shykes** [comment]
You were clear the first time, I do understand how the request is processed. The resolver for eg. `Container.directory`, `Directory.directory` is not lazy. I'm saying, OK just make it lazy then :) It will match the current developer expectation, and will solve the problem at hand in a simple and elegant way.

You mentioned that making those resolvers lazy would make error messages less useful, but I don't understand how. Can you give an example?

**Comment by helderco** [comment]
> You mentioned that making those resolvers lazy would make error messages less useful, but I don't understand how. Can you give an example?

This validation kind of makes sense to me in how it can improve error messages:

https://github.com/dagger/dagger/blob/de39db1dac4728888a65e05ee131b121f958e298/core/container.go#L735-L740

One example that’s not directly related to this, but is similar, was when I was adding the global connection to Python and started with making the connection lazy. If there was an error in that connection, it would happen in the first field that needed to make a request, usually an obscure `id` that the SDK hides from you. Took me a while to realize what was happening because the error was very obscure so I decided not to push for lazy connections after that.

Just saying that to explain kind of why the validation makes sense to me, but this is different so let’s question the validation here because it does force it to unlazy in the resolver and that has it’s own downsides.

I tracked it down to this comment from @sipsma in:
- https://github.com/dagger/dagger/pull/3217#discussion_r987475016

> If you provide a dir that doesn't exist, this won't fail, right? It would only fail at some later time when you tried to do something with the Directory.
>
> If so: it's tempting to add a check here that the dir actually exists (via gateway ReadDir or similar). There technically is a downside to that in that it will cause some buildkit refs to be "unlazied" (i.e. buildkit defers pulling an image until its really needed locally, same for calculating mergeops).
>
> I'm inclined to not care about that performance optimization at the moment because the surprise factor of the error messages you might later get if the dir provided here doesn't exist is probably quite high.

That was on Oct 5, 2022 (when we were getting ready to release the Go SDK) which was a long time ago and seems like more a precaution.

I can remove the validation but it'll take me time to hunt down various ways where it can fail in order to compare how it affects the user experience. Could be negligible in some cases, a breaking change in others, or just more confusing. It’s not something we can take back as the `.Exists` field seizes to work so that’s why I’m a bit cautious.

I can confirm that `dag.Directory().Directory("/foo").Sync(ctx)` doesn’t return an error but I can add the validation to `sync`. I’ll try it for size when I can (soon).

@sipsma and @vito, do you have anything to add?

**Comment by rajatjindal** [comment]
I have started to look at this issue. could we assign this to me please. thank you.

**Comment by jedevc** [comment]
@rajatjindal can do - I would probably start with https://github.com/dagger/dagger/issues/3764 to familiarize yourself with the file API, but up to you.

Probably also worth discussing with @helderco before starting, just to check what the desired outcome looks like, I'm not sure if anything has changed in the last few months that might change what we want.

**Comment by rajatjindal** [comment]
Hi @jedevc, sure, I will take a look at that first. thank you for your suggestion.

**Comment by rajatjindal** [comment]
Hi @helderco, I am currently going through the comments here and on #3764 (and related tickets) to build context on what needs to change and where.

I have a question regarding your comment above:

> I can confirm that dag.Directory().Directory("/foo").Sync(ctx) doesn’t return an error, for example, but I can add the validation to sync in this case.

I tried adding a small testcase for verifying current laziness of `Directory`, and noticed that it fails in above stated scenario as well:

```
func (DirectorySuite) TestDirectoryLaziness(ctx context.Context, t *testctx.T) {
	c := connect(ctx, t)

	_, err := c.Directory().Directory("/dont-exist").Sync(ctx)
	require.Error(t, err)

	_, err = c.Container().Directory("/dont-exist").Sync(ctx)
	require.Error(t, err)

	_, err = c.Container().Directory("/dont-exist").Entries(ctx)
	require.Error(t, err)
}
```

all the above functions calls return `/dont-exist: no such file or directory` error. 

could you please help me confirm if that is expected or I am doing something incorrectly here. 

thanks

**Comment by helderco** [comment]
> Hi @helderco, I am currently going through the comments here and on #3764 (and related tickets) to build context on what needs to change and where.

I don't think we'll want to move forward with:
- #3764

This `Directory.exists` and `File.exists` supercedes the need for it. Whatever else that `Stat` could provide can intead be its own field in `Directory` or  `File`. See this comment above: https://github.com/dagger/dagger/issues/6713#issuecomment-2094536129.

---

> I have a question regarding your comment above:
> 
> > I can confirm that dag.Directory().Directory("/foo").Sync(ctx) doesn’t return an error, for example, but I can add the validation to sync in this case.
> 
> [...]
>
> all the above functions calls return `/dont-exist: no such file or directory` error.
> 
> could you please help me confirm if that is expected or I am doing something incorrectly here.

Yes, the current behavior is that `Directory("/foo")` is validating the existence of the directory, not even getting a chance to execute `sync`. In the client, the call to `Sync(ctx)` just produces a query and makes the request, but in the server side, that function doesn't even run.

The point was, for you to create a new `dag.Directory().Directory("/foo").Exists(ctx)`, it'll fall under the same problem where `Exists` doesn't get a chance to run in the server because of the same validation in `Directory("/foo")`.

So first thing you need to do is to remove that validation in `Directory` (see snippet in https://github.com/dagger/dagger/issues/6713#issuecomment-2097115794). Now, if you do `dag.Directory().Directory("/foo").Sync(ctx)`, it shouldn't fail. That's because `sync` evaluates the pipeline, but there's nothing that needs to use the `dag.Directory().Directory("/foo")` directory, and so without that validation in place, it won't fail.

The implication here is that it can break users that may rely on this early validation that exists today in similar cases, so they'll have to explicitly use `.Exists(ctx)` in those situations. In most cases, the implication is that it can make the error message more confusing for debugging purposes, since the error should happen when the `Directory` is used (as an input to another function), rather than when it's evaluated. However, the early validation does break laziness, so it's expected that fixing that may improve performance in some cases.

In conclusion, it's useful to try and move that validation from `Directory.directory` (etc) into `Directory.exists` and evaluate how it affects the user experience. The hope is that the benefit of an explicit `Directory.exists` outweighs the early validation in `Directory.directory`.

**Comment by rajatjindal** [comment]
Hi @helderco , I have started work on this and started making some changes in a [WIP] branch [here](https://github.com/dagger/dagger/compare/main...rajatjindal:dagger:add-exists-fn?expand=1).

I have a couple of questions:

1. could you please let me know what would be the expected result in following pipeline. with the lazyness changes, it returns an ID (because we removed validation from `Directory` function). If we want this to return an error, could you please point me to where I would need to make change for making validation work during an `ID` function:

```go
id, err := c.Container().
            From("alpine:latest").
            Directory("/dont-exist").ID(ctx)
```

2. similarly for following scenario, it currently does not return any error because in `Evaluate` [function](https://github.com/dagger/dagger/blob/742841f49b208ece1a4f9a105939319e1eee7a16/core/directory.go#L156), we return nil without any error if `dir.LLB` is nil. 

Most of the other `TestDirectory` and `TestContainer` tests work beside these two scenarios.

Thanks
Rajat Jindal

**Comment by helderco** [comment]
> 1. could you please let me know what would be the expected result in following pipeline. with the lazyness changes, it returns an ID (because we removed validation from `Directory` function). If we want this to return an error, could you please point me to where I would need to make change for making validation work during an `ID` function:
> 
> ```go
> id, err := c.Container().
>             From("alpine:latest").
>             Directory("/dont-exist").ID(ctx)
> ```

To preserve the lazyiness you don't validate on `ID`. It should either error where it's used or you explicitly call `Exists` to validate.

I can see users needing a lot of `Exists` in all sorts of places though because of this. There must be a better way.

> 2. similarly for following scenario, it currently does not return any error because in `Evaluate` [function](https://github.com/dagger/dagger/blob/742841f49b208ece1a4f9a105939319e1eee7a16/core/directory.go#L156), we return nil without any error if `dir.LLB` is nil.
> 
> ```go
> _, err := c.Directory().Directory("/foo").Sync(ctx)
> require.Error(t, err)
> require.Contains(t, err.Error(), "no such file or directory")
> ```

This is probably the same as `ID`, however I wonder if it could validate implicitly. 

Let's see, what if you use it to validate an `ID`?
```go
dir, err := c.Directory().Directory("/doesnt-exist").Sync(ctx)
if err != nil {
    return err
}
ctr.WithDirectory("/sub", dir)  // The SDK requests the ID here before making the query
```

Isn't this almost the same (needs error handling)?
```go
dir := c.Directory().Directory("/doesnt-exist")
ok, err := dir.Exists(ctx)
if err != nil {
    return err
}
if !ok {
    return fmt.Errorf("my own error message")
}
ctr.WithDirectory("/sub", dir)
```

Small benefit, but still a convenience. Perhaps it would be better to do this instead:
```go
dir := c.Directory().WithNewFile("/a/b/my-file", "")

ok, err := dir.Directory("/a").Exists(ctx)
if err != nil {
    return err
}
if !ok {
    return fmt.Errorf("no cookie")
}

ctr.WithDirectory(
    "/sub",
    dir.Directory("/a/c", DirectoryDirectoryOpts{
        Check: true,  // equivalent to today, just explicitly enabled
    }),
)
```

Or even better, flip the default. Basically disable if you need `Exists` for more advanced logic, but default to validating:
```go
dir := c.Directory().WithNewFile("/a/b/my-file", "")

ok, err := dir.
    Directory("/a", DirectoryDirectoryOpts{
        SkipValidation: true,
    }).
    Exists(ctx)
if err != nil {
    return err
}
if !ok {
    return fmt.Errorf("no cookie")
}

ctr.WithDirectory(
    "/sub",
    dir.Directory("/a/c"),  // works like today, fails at this point in the pipeline
)
```

I think the last example is the better UX because if you need to rely on `Exists`, user pipelines will become much more verbose. So it allows the fine grained control when needed but still allows the simplest default for most cases. 

However, the downside is that a user may attempt this:
```go
ok, err := dir.Directory("/a").Exists(ctx)
if err != nil {
    return err
}
if !ok {
    return fmt.Errorf("no cookie")  // ❌ never happens!
}
```

And not realize that it won’t work without the `SkipValidation: true` option.

TL;DR

In light of this, **I think the most sensible thing is to add the `Check: true` option**. This way it also validates on `ID`, `Sync`, etc… if you want simple validation without having to call `Exists`.

**Comment by shykes** [comment]
Can someone please give me a concrete example of how making the existence check lazy actually makes the DX worse? I see that assumption being made everywhere, with lots of discussions on how to mitigate it. But are we sure there is really something to mitigate? Concrete examples would really help here.

**Comment by helderco** [comment]
Yeah, I think part of the work here is exploring just that, in order to compare the DX. See what error messages you get when you use a directory as input in several different cases.

**Comment by rajatjindal** [comment]
Hi @shykes, @helderco 

Following are some of the testcases for checking laziness error. 

The testcases works as expected on `main` branch because we have validation in `directory.Directory` and `container.Directory` methods. But it fails with different errors once we remove validation from both these functions. 

```go

func (DirectorySuite) TestDirectoryLazinessError(ctx context.Context, t *testctx.T) {
	c := connect(ctx, t)

	t.Run("container", func(ctx context.Context, t *testctx.T) {
		t.Run("tc1", func(ctx context.Context, t *testctx.T) {
			// without validation, it return error "input: container.from.directory.asModule.initialize resolve: module name and SDK must be set\n"
			err := c.Container().From("alpine:latest").Directory("/doesnt-exist").AsModule().Initialize().Serve(ctx)
			require.ErrorContains(t, err, "lstat /doesnt-exist: no such file or directory")
		})

		t.Run("tc2", func(ctx context.Context, t *testctx.T) {
			// without validation, it returns "input: container.from.directory.diff.entries resolve: resolve : lstat /tmp/buildkit-mount2166899413/doesnt-exist: no such file or directory\n"
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").Diff(c.Directory().Directory("/doesnt-exist")).Entries(ctx)
			require.ErrorContains(t, err, "/doesnt-exist: no such file or directory")
		})

		t.Run("tc3", func(ctx context.Context, t *testctx.T) {
			c := connect(ctx, t)
			// without validation, it returns "input: container.from.directory.file resolve: lstat /doesnt-exist/foo.txt: no such file or directory\n"
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").File("foo.txt").Contents(ctx)
			require.ErrorContains(t, err, "lstat /doesnt-exist: no such file or directory")
		})

		t.Run("tc4", func(ctx context.Context, t *testctx.T) {
			// without validation, it returns "input: container.from.directory.entries resolve: resolve : lstat /tmp/buildkit-mount3306785835/doesnt-exist: no such file or directory\n"
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").Entries(ctx)
			require.ErrorContains(t, err, "lstat /doesnt-exist: no such file or directory")
		})

		t.Run("tc5", func(ctx context.Context, t *testctx.T) {
			// without validation, this does not return any error
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").Sync(ctx)
			require.ErrorContains(t, err, "lstat /doesnt-exist: no such file or directory")
		})

		t.Run("tc6", func(ctx context.Context, t *testctx.T) {
			// without validation, it returns error -> "input: container.from.directory.glob resolve: resolve : lstat /tmp/buildkit-mount2890696365/doesnt-exist: no such file or directory\n"
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").Glob(ctx, "*.*")
			require.ErrorContains(t, err, "lstat /doesnt-exist: no such file or directory")
		})
	})
}
```

It does feel like removing the validation by default will return error msgs which are not user friendly. and may also result in unexpected results for already published modules. e.g. `tc6` in above list does not return any error after validation is removed.

Please let me know what you think and how we should proceed on this.

**Comment by helderco** [comment]
Oh, I suppose the `require.ErrorContains` is the "before" error, and the comments are the "after".

**Comment by helderco** [comment]
There's no examples with `WithDirectory`, for example. Would also be helpful to diversify a bit more and look into more use cases.

**Comment by rajatjindal** [comment]
Hi @helderco, 

following is the updated list with `WithDirectory` example. Few things to note:

- I have captured the complete error msg now
- The enabled assertion is what happens when validation is enabled
- The commented out assertion is what happens when validations is disabled.
- This is the output with disabling validations only in Directory.Directory and Container.Directory functions and no other changes done elsewhere.
- These changes are pushed in a WIP branch [here](https://github.com/dagger/dagger/compare/main...rajatjindal:dagger:laziness-test?expand=1)
- 
I will continue to add more testcases to this list to cover more usecases tomorrow.

```go

func (DirectorySuite) TestDirectoryLazinessError(ctx context.Context, t *testctx.T) {
	c := connect(ctx, t)

	t.Run("container", func(ctx context.Context, t *testctx.T) {
		t.Run("tc1", func(ctx context.Context, t *testctx.T) {
			err := c.Container().From("alpine:latest").Directory("/doesnt-exist").AsModule().Initialize().Serve(ctx)
			require.EqualError(t, err, "input: container.from.directory resolve: lstat /doesnt-exist: no such file or directory\n")
			// require.EqualError(t, err, "input: container.from.directory.asModule.initialize resolve: module name and SDK must be set\n")
		})

		t.Run("tc2", func(ctx context.Context, t *testctx.T) {
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").Diff(c.Directory().Directory("/doesnt-exist")).Entries(ctx)
			require.EqualError(t, err, "input: directory.directory resolve: /doesnt-exist: no such file or directory\n")
			// require.EqualError(t, err, "input: container.from.directory.diff.entries resolve: resolve : lstat /tmp/buildkit-mount3884148634/doesnt-exist: no such file or directory\n"
		})

		t.Run("tc3", func(ctx context.Context, t *testctx.T) {
			c := connect(ctx, t)
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").File("foo.txt").Contents(ctx)
			require.EqualError(t, err, "input: container.from.directory resolve: lstat /doesnt-exist: no such file or directory\n")
			// require.EqualError(t, err, "input: container.from.directory.file resolve: lstat /doesnt-exist/foo.txt: no such file or directory\n")
		})

		t.Run("tc4", func(ctx context.Context, t *testctx.T) {
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").Entries(ctx)
			require.EqualError(t, err, "input: container.from.directory resolve: lstat /doesnt-exist: no such file or directory\n")
			// require.EqualError(t, err, "input: container.from.directory.entries resolve: resolve : lstat /tmp/buildkit-mount3859364930/doesnt-exist: no such file or directory\n")
		})

		t.Run("tc5", func(ctx context.Context, t *testctx.T) {
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").Sync(ctx)
			require.EqualError(t, err, "input: container.from.directory resolve: lstat /doesnt-exist: no such file or directory\n")
			// require.NoError(t, err)
		})

		t.Run("tc6", func(ctx context.Context, t *testctx.T) {
			_, err := c.Container().From("alpine:latest").Directory("/doesnt-exist").Glob(ctx, "*.*")
			require.EqualError(t, err, "input: container.from.directory resolve: lstat /doesnt-exist: no such file or directory\n")
			// require.EqualError(t, err, "input: container.from.directory.glob resolve: resolve : lstat /tmp/buildkit-mount426363861/doesnt-exist: no such file or directory\n")
		})

		t.Run("tc7", func(ctx context.Context, t *testctx.T) {
			dir := c.Container().From("alpine:latest").Directory("/doesnt-exist")

			_, err := c.
				Container().
				From("alpine:latest").
				WithDirectory("/mounted-foo", dir).
				Directory("/mounted-foo").
				Glob(ctx, "*.*")

			require.EqualError(t, err, "input: container.from.directory resolve: lstat /doesnt-exist: no such file or directory\n")
			// require.EqualError(t, err, "input: container.from.withDirectory.directory.glob resolve: failed to compute cache key: failed to calculate checksum of ref n9lagq287siw2a6os7nzr7ix7::hnrt1bom67rf2upvpvhamp6xb: \"/doesnt-exist\": not found\n")
		})

		t.Run("tc8", func(ctx context.Context, t *testctx.T) {
			_, err := c.
				Container().
				From("alpine:latest").
				WithDirectory("/mounted-foo", c.Container().From("alpine:latest").Directory("/doesnt-exist")).
				Directory("/mounted-foo").
				Glob(ctx, "*.*")

			require.EqualError(t, err, "input: container.from.directory resolve: lstat /doesnt-exist: no such file or directory\n")
			// require.EqualError(t, err, "input: container.from.withDirectory.directory.glob resolve: failed to compute cache key: failed to calculate checksum of ref n9lagq287siw2a6os7nzr7ix7::hnrt1bom67rf2upvpvhamp6xb: \"/doesnt-exist\": not found\n")
		})

		t.Run("tc9", func(ctx context.Context, t *testctx.T) {
			_, err := c.
				Container().
				From("alpine:latest").
				WithMountedDirectory("/mounted-foo", c.Container().From("alpine:latest").Directory("/doesnt-exist")).
				Directory("/mounted-foo").
				Glob(ctx, "*.*")

			require.EqualError(t, err, "input: container.from.directory resolve: lstat /doesnt-exist: no such file or directory\n")
			// require.EqualError(t, err, "input: container.from.withMountedDirectory.directory.glob resolve: resolve : lstat /tmp/buildkit-mount1338184602/doesnt-exist: no such file or directory\n")
		})
	})
}
```

**Comment by rajatjindal** [comment]
Hi @helderco, in context of removing validation from `container.Directory` and `directory.Directory`, is there any additional testcase you would like to consider? I added a couple more scenarios, but it seems like I am repeating the testcases now.

**Comment by shykes** [comment]
@rajatjindal thanks for sharing the error messages before/after the change, that is helpful. It's true that some of those messages are less clear if we make the change. Most of them still have the directory name + some variation of "this doesn't exist" so it might be OK. But it's definitely a downgrade in terms of error message clarity and brevity.

With that in mind, maybe we should consider either:

1) @helderco 's suggestion of making the new behavior opt-in (`CheckExists` optional argument or equivalent)
2) Change approach completely and add a `path` argument to `Exists`, eg. `Directory.Exists("./not-exist")`
3) Go even further in that direction and implement a `Stat` which would be a more general version of that.

I realize some of these options have been suggested before :) This all hinges on how bad we think the error message downgrade will be. Which is why getting real-world examples is so important.
